### DISCUSSION TITLE
Clang 和 GCC 下 对构造函数使用SFINAE 的问题

### DISCUSSION THREAD

**Discussion by AzrBrk** [Mq-b/Modern-Cpp-templates-tutorial#18] [Q&A] [answered]
# 关于对复制和移动构造函数应用SFINAE的问题

## 起因

最初的目的是在我所编写的类中控制移动和构造函数的使用，以确保在不满足特定条件时不调用定义的移动函数，而是调用复制构造函数。我通过简短的代码尝试复现了这个问题。需要注意的是，在MSVC编译器中，这个问题被忽略了，编译通过。

## 复现

以下是相关的代码：

```cpp
#include <type_traits>

template <class T>
class A {
private:
    T a{};

public:
    A() {}

    template <class En = std::enable_if_t<std::is_same_v<T, int>>>
    A(const A& aa) {
        a = aa.a;
    }

    template <class En = std::enable_if_t<std::is_same_v<T, int>>>
    A(A&& aa) {
        a = aa.a;
    }
};

int main() {
    A<int> b{};      // is_same_v<int, int>为true，正常
    A<double> bb{};  // is_same_v<int, int>为false，出错
}
```

然而，Clang 编译器给出了以下错误信息：

```cpp
error: no type named 'type' in 'std::enable_if<false>'; 'enable_if' cannot be used to disable this declaration
```

## 问题

因此，是否能够在复制或者构造函数中应用 SFINAE 机制呢？

## 链接

[查看示例](https://gcc.godbolt.org/z/jcq7aKE5b)

**Comment by 16bit-ykiko** [comment]
> # 关于对构造函数应用SFINAE的问题
> ## 起因
> 我原来的本意是在我写的类中控制移动和构造函数的使用, 使得在不满足某些条件时不要调用定义的移动函数, 而去调用复制构造函数. 我用简短的代码复现了一下. 注意, 此问题在MSVC中被无视了,编译通过.
> 
> ## 复现
> 代码如下:
> 
> ```c++
> #include<type_traits>
> template<class T>
> class A{
> private:
>     T a{};
> public:
>     A(){}
> 
>     template<class En = std::enable_if_t<std::is_same_v<T, int>>>
>     A(const A & aa){
>         a = aa.a;
>     }
> 
>     template<class En = std::enable_if_t<std::is_same_v<T, int>>>
>     A(A && aa){
>         a = aa.a;
>     }
> 
> };
> 
> int main(){
>     A<int> b{};//is_same_v<int, int> true,OK
>     A<double> bb{};//is_same_v<int, int> false, Error
> }
> ```
> 
> clang 给出如下报错信息:
> 
> * error: no type named 'type' in 'std::enable_if'; 'enable_if' cannot be used to disable this declaration
> 
> ## 问题
> 所以到底能不能对构造函数使用SFINAE的机制？
> 
> ## 链接
> https://gcc.godbolt.org/z/jcq7aKE5b

你这里构造函数都不是模板函数，你的 SFINAE 是对谁用？

**Reply by AzrBrk** [reply]
嗯复制构造和移动构造

**Comment by 16bit-ykiko** [comment]
注意实例化的时机，进行延迟实例化即可

https://gcc.godbolt.org/z/rsbcr6v7z

**Reply by AzrBrk** [reply]
可以的,我又试了普通函数,好像只要不使用延迟实例化都不行:

https://gcc.godbolt.org/z/q44Mb383e

**Reply by 16bit-ykiko** [reply]
注意到模板实例化中的 immediate context，只有在immediate context 中替换失败才不会导致 hard error，而对于类模板而言，成员函数的声明并不属于 immediate context。

**Reply by AzrBrk** [reply]
我好像有点明白了,SFINAE的条件是，只有找到一个可以被实例化的版本， 编译器会忽略掉这个错误的实例化，否则就会出错，应该是因为这样的原因？所以一开始我的写法是错误的，但是msvc不知道为什么通过了。

**Reply by Mq-b** [reply]
**只有在函数类型或其模板形参类型或其 explicit 说明符 (C++20 起)的立即语境中的类型与表达式中的失败，才是 SFINAE 错误**。如果对代换后的类型/表达式的求值导致副作用，例如实例化某模板特化、生成某隐式定义的成员函数等，那么这些副作用中的错误都被当做硬错误。

**Reply by Mq-b** [reply]
视频和教案里其实提了，不过侧重点到了“产生副作用”上了，给了个示例。

**Reply by Mq-b** [reply]
> 但是msvc不知道为什么通过了。

不清楚。

**Reply by frederick-vs-ja** [reply]
MSVC 一直有函数模板中一些错误在实例化时才触发的问题。而这个例子很巧地（在改正后）不会触发函数模板特化的实例化。

**Comment by Mq-b** [comment]
@frederick-vs-ja

**Comment by frederick-vs-ja** [comment, marked as answer]
首先问题描述本身存在一些错误：**复制构造函数与移动构造函数都不能是函数模板**。
所以样例代码中涉及的构造函数都不是复制或移动构造函数。而且复制构造函数和移动构造函数仍然会被隐式声明（假设模板头被改正），从而题主引入的函数模板不会得到选择。

如果要对复制或移动特殊成员函数使用 SFINAE，可以这么做：
- 使用 C++20 `requires`，或
- 引入对应的特殊成员函数条件性 `= delete;` 或 `= default;` 的**基类**或**非静态数据成员**，标准库中 `std::optional` 与 `std::variant` 的实现就使用各种专门定制的基类（因为需要用 C++17 语核实现）。

----
我们可以记住一条规则：**参与传统 SFINAE 的模板形参一定要是在当前模板中引入的**。
外围的模板形参不再参与推导或替换，从而会导致硬错误。

----
不过，如果题主的意图不是“使用复制或移动构造函数”，而是“使用函数模板，但使之能用于复制或移动构造”的话，可以这么做。（[Godbolt link](https://gcc.godbolt.org/z/PzsjGbYKn)）

```C++
#include <type_traits>
#include <utility>

template<class T>
class A{
private:
    T a{};

    A(const volatile A&) = delete; // This is a copy ctor!

public:
    A() = default;

    template<class U = T, class = std::enable_if_t<std::is_same_v<U, int>>>
    A(const A & aa) : a(aa.a) {}

    template<class U = T, class = std::enable_if_t<std::is_same_v<U, int>>>
    A(A && aa) noexcept : a(std::move(aa.a)) {}
};
```

此处我们显式声明了一个复制构造函数，它会抑制复制构造函数与移动构造函数的隐式声明。从而下面两个函数模板会得到选择，并被用于复制或移动构造。

**Reply by Mq-b** [reply]
聚聚无敌，聚聚：“***为什么好像除了我大家都假装不知道函数模板不会是复制/移动构造函数？***”

**Reply by Mq-b** [reply]
> ***参与传统 SFINAE 的模板形参一定要是在当前模板中引入的。***

好好好，其实说的就是***立即语境***。
